standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/DatabaseProduct.java

```diff
@@ -546,41 +553,48 @@ public boolean supportsGetGeneratedKeys() throws MetaException {
     }
   }
 
-  public boolean isDuplicateKeyError(SQLException ex) {
+  public boolean isDuplicateKeyError(Throwable ex) {
+    while (ex != null && !(ex instanceof SQLException) ) {
+      ex = ex.getCause();
+    }
+    if (ex == null) {
+      throw new IllegalArgumentException("No SQLException found in the exception chain!");
```

Comment:
This part is the same in isTableNotExistsError. Can you move it to a separate method?

Reply:
Created a method in TxnUtils.